yarpl/include/yarpl/flowable/EmitterFlowable.h

```diff
@@ -26,25 +26,20 @@ class EmiterBase : public virtual Refcounted {
  * This is synchronous: the emit calls are triggered within the context
  * of a request(n) call.
  */
-template<typename T>
-class EmiterSubscription : private Subscription, private Subscriber<T> {
+template <typename T>
+class EmiterSubscription : public virtual Subscription,
+                           public virtual Subscriber<T> {
   constexpr static auto kCanceled = credits::kCanceled;
   constexpr static auto kNoFlowControl = credits::kNoFlowControl;
 
  public:
   EmiterSubscription(
-      Reference <EmiterBase<T>> emiter,
-      Reference <Subscriber<T>> subscriber)
-      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {
-    // We expect to be heap-allocated; until this subscription finishes
-    // (is canceled; completes; error's out), hold a reference so we are
-    // not deallocated (by the subscriber).
-    Refcounted::incRef(*this);
-    subscriber_->onSubscribe(get_ref<Subscription>(this));
-  }
+      Reference<EmiterBase<T>> emiter,
+      Reference<Subscriber<T>> subscriber)
+      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {}
 
-  virtual ~EmiterSubscription() {
-    subscriber_.reset();
+  void init() {
+    subscriber_->onSubscribe(get_ref<Subscription>(this));
```

Comment:
The problem might be that, which is similar to shared_ptr's.
http://en.cppreference.com/w/cpp/memory/enable_shared_from_this
Check this link. In there it says that, if we call `shared_from_this`, from an instance which is not a shared_ptr, it causes undefined behavior.

So I need to ask, if we call get_ref<T>(this), but if there is no Reference<T> outside of `this`, then if we call get_ref - increase count to 1, when we lose this ref, count goes to 0, which deletes the object.

So the problem is somewhere we create an instance of the Emitter without a reference! Which is the constructor of the Emitter already. Namely even if we have Reference<Emitter>, at the constructor of the Emitter, there is nothing that is incrementing the count of the reference. 

I believe the inc_count and dec_count in the code in other places were specifically for this purpose. So we need inc_cnt / dec_cnt in here too. Perhaps we need to revert the old update that has removed these two calls from another class.